plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/resolve/KSerializerDescriptorResolver.kt

```diff
@@ -366,13 +366,20 @@ object KSerializerDescriptorResolver {
 
         val args = mutableListOf<ValueParameterDescriptor>()
         var i = 0
+
+        val targ = TypeParameterDescriptorImpl.createWithDefaultBound(f, Annotations.EMPTY, false, Variance.INVARIANT, Name.identifier("T"), 0)
```

Comment:
It would be perfect to take type arguments from `thisClass` instead of creating them in-place, because they should match – to operate on `Data<E,V>` we need `KSerializer<E>` and `KSerializer<V>`. 
However, this might be tricky and requires knowledge of compiler API; to bypass annotation processors/bytecode validators, AFAIK, it is sufficient just to declare them in-place.